src/cart/crt_internal_types.h

```diff
@@ -166,6 +166,160 @@ struct crt_event_cb_priv {
+	ENV_STR(D_FI_CONFIG)                                                                       \
+	ENV_STR(D_INTERFACE)                                                                       \
+	ENV_STR(D_LOG_FILE)                                                                        \
+	ENV_STR(D_LOG_FILE_APPEND_PID)                                                             \
+	ENV_STR(D_LOG_FILE_APPEND_RANK)                                                            \
+	ENV_STR(D_LOG_FLUSH)                                                                       \
+	ENV_STR(D_LOG_MASK)                                                                        \
+	ENV_STR(D_LOG_SIZE)                                                                        \
+	ENV(D_LOG_STDERR_IN_LOG)                                                                   \
+	ENV(D_POLL_TIMEOUT)                                                                        \
+	ENV_STR(D_PORT)                                                                            \
+	ENV(D_PORT_AUTO_ADJUST)                                                                    \
+	ENV(D_POST_INCR)                                                                           \
+	ENV(D_POST_INIT)                                                                           \
+	ENV_STR(D_PROVIDER)                                                                        \
+	ENV_STR_NO_PRINT(D_PROVIDER_AUTH_KEY)                                                      \
+	ENV(D_QUOTA_RPCS)                                                                          \
+	ENV(FI_OFI_RXM_USE_SRX)                                                                    \
+	ENV(FI_UNIVERSE_SIZE)                                                                      \
+	ENV(SWIM_PING_TIMEOUT)                                                                     \
+	ENV(SWIM_PROTOCOL_PERIOD_LEN)                                                              \
+	ENV(SWIM_SUSPECT_TIMEOUT)                                                                  \
+	ENV_STR(UCX_IB_FORK_INIT)
+
+/* uint env */
+#define ENV(x)                                                                                     \
+	unsigned int _##x;                                                                         \
+	int          _rc_##x;                                                                      \
+	int          _no_print_##x;
+
+/* char* env */
+#define ENV_STR(x)                                                                                 \
+	char *_##x;                                                                                \
+	int   _rc_##x;                                                                             \
+	int   _no_print_##x;
+
+#define ENV_STR_NO_PRINT(x) ENV_STR(x)
+
+struct crt_envs_t {
+	CRT_ENV_LIST;
+	bool inited;
+};
+
+#undef ENV
+#undef ENV_STR
+#undef ENV_STR_NO_PRINT
+
+extern struct crt_envs_t g_envs;
```

Comment:
ideally you'd want to use `__attribute__ ((visibility ("hidden")))` for all global private declarations so that they remain hidden from public users.

Reply:
hmm.. i remember before we had something in scons to make everything hidden by default, but i cant find it. Does it ring any bells @jolivier23 ?

Reply:
anyway, i tried adding visibility hidden here and it has issues linking numerous files:
`build/release/gcc/src/cart/utils/s_crt_utils.os: In function `write_completion_file':
/home/aaoganez/github/daos_master/src/cart/utils/crt_utils.c:104: undefined reference to `crt_genvs'
/home/aaoganez/github/daos_master/src/cart/utils/crt_utils.c:104: undefined reference to `crt_genvs'
/home/aaoganez/github/daos_master/src/cart/utils/crt_utils.c:104: undefined reference to `crt_genvs'
`

Reply:
Ah ok well if you use it in crt_utils then it's not really private 😄

Reply:
Yeah, we did this in CPPR and IOF via an option to the linker and then defined a PUBLIC define to mark things visible

Reply:
Looks like it carried over to src/client/dfuse/SConstruct where we set -fvisibility=hidden and then we use DFUSE_PUBLIC to mark the visible routines.